# Finding the Best Markets to Advertise In

This time we're working for an e-learning company that offers courses on programming. Most of our courses are on web and mobile development, but we also cover many other domains, like data science, game development, etc. We want to promote our product and we'd like to invest some money in advertisement. Our goal in this project is to find out the two best markets to advertise our product in.

To reach our goal, we could organize surveys for a couple of different markets to find out which would the best choices for advertising. This is very costly, however, and it's a good call to explore cheaper options first.

We can try to search existing data that might be relevant for our purpose. One good candidate is the data from [freeCodeCamp's 2017 New Coder Survey](https://medium.freecodecamp.org/we-asked-20-000-people-who-they-are-and-how-theyre-learning-to-code-fff5d668969). [freeCodeCamp](https://www.freecodecamp.org/) is a free e-learning platform that offers courses on web development. Because they run [a popular Medium publication](https://medium.freecodecamp.org/) (over 400,000 followers), their survey attracted new coders with varying interests (not only web development), which is ideal for the purpose of our analysis.

The survey data is publicly available in [this GitHub repository](https://github.com/freeCodeCamp/2017-new-coder-survey).

In this project, I would like to use plotly for interactive visualization for our convenience.

First, let's import and explore the data!


## Exploring The Data

In [1]:
import numpy as np
import pandas as pd
import plotly.plotly as py
import plotly.graph_objs as go
import cufflinks as cf

import warnings
warnings.filterwarnings('ignore')

data = pd.read_csv('2017-fCC-New-Coders-Survey-Data.csv')

data.shape

(18175, 136)

In [2]:
data.describe()

Age  AttendedBootcamp  BootcampFinish  BootcampLoanYesNo  \
count  15367.000000      17709.000000     1069.000000        1079.000000   
mean      27.691872          0.062002        0.699719           0.305839   
std        8.559239          0.241167        0.458594           0.460975   
min        0.000000          0.000000        0.000000           0.000000   
25%       22.000000          0.000000        0.000000           0.000000   
50%       26.000000          0.000000        1.000000           0.000000   
75%       32.000000          0.000000        1.000000           1.000000   
max       90.000000          1.000000        1.000000           1.000000   

       BootcampRecommend  ChildrenNumber  CodeEventConferences  \
count        1073.000000     2314.000000                1609.0   
mean            0.818267        1.832325                   1.0   
std             0.385805        0.972813                   0.0   
min             0.000000        1.000000                   1.0   
25%             1.000000        1.000000                   1.0   
50%             1.000000        2.000000                   1.0   
75%             1.000000        2.000000                   1.0   
max             1.000000        9.000000                   1.0   

       CodeEventDjangoGirls  CodeEventFCC  CodeEventGameJam  \
count                 165.0        1708.0             290.0   
mean                    1.0           1.0               1.0   
std                     0.0           0.0               0.0   
min                     1.0           1.0               1.0   
25%                     1.0           1.0               1.0   
50%                     1.0           1.0               1.0   
75%                     1.0           1.0               1.0   
max                     1.0           1.0               1.0   

              ...           YouTubeEngineeredTruth  YouTubeFCC  \
count         ...                            993.0      6036.0   
mean          ...                              1.0         1.0   
std           ...                              0.0         0.0   
min           ...                              1.0         1.0   
25%           ...                              1.0         1.0   
50%           ...                              1.0         1.0   
75%           ...                              1.0         1.0   
max           ...                              1.0         1.0   

       YouTubeFunFunFunction  YouTubeGoogleDev  YouTubeLearnCode  \
count                 1261.0            3539.0            2662.0   
mean                     1.0               1.0               1.0   
std                      0.0               0.0               0.0   
min                      1.0               1.0               1.0   
25%                      1.0               1.0               1.0   
50%                      1.0               1.0               1.0   
75%                      1.0               1.0               1.0   
max                      1.0               1.0               1.0   

       YouTubeLevelUpTuts  YouTubeMIT  YouTubeMozillaHacks  \
count              1396.0      3327.0                622.0   
mean                  1.0         1.0                  1.0   
std                   0.0         0.0                  0.0   
min                   1.0         1.0                  1.0   
25%                   1.0         1.0                  1.0   
50%                   1.0         1.0                  1.0   
75%                   1.0         1.0                  1.0   
max                   1.0         1.0                  1.0   

       YouTubeSimplilearn  YouTubeTheNewBoston  
count               201.0               2960.0  
mean                  1.0                  1.0  
std                   0.0                  0.0  
min                   1.0                  1.0  
25%                   1.0                  1.0  
50%                   1.0                  1.0  
75%                   1.0                  1.0  
max      

In [3]:
data.head()

Age  AttendedBootcamp  BootcampFinish  BootcampLoanYesNo BootcampName  \
0  27.0               0.0             NaN                NaN          NaN   
1  34.0               0.0             NaN                NaN          NaN   
2  21.0               0.0             NaN                NaN          NaN   
3  26.0               0.0             NaN                NaN          NaN   
4  20.0               0.0             NaN                NaN          NaN   

   BootcampRecommend  ChildrenNumber                 CityPopulation  \
0                NaN             NaN            more than 1 million   
1                NaN             NaN              less than 100,000   
2                NaN             NaN            more than 1 million   
3                NaN             NaN  between 100,000 and 1 million   
4                NaN             NaN  between 100,000 and 1 million   

   CodeEventConferences  CodeEventDjangoGirls         ...           \
0                   NaN                   NaN         ...            
1                   NaN                   NaN         ...            
2                   NaN                   NaN         ...            
3                   NaN                   NaN         ...            
4                   NaN                   NaN         ...            

   YouTubeFCC  YouTubeFunFunFunction  YouTubeGoogleDev  YouTubeLearnCode  \
0         NaN                    NaN               NaN               NaN   
1         1.0                    NaN               NaN               NaN   
2         NaN                    NaN               NaN               1.0   
3         1.0                    1.0               NaN               NaN   
4         NaN                    NaN               NaN               NaN   

   YouTubeLevelUpTuts  YouTubeMIT  YouTubeMozillaHacks YouTubeOther  \
0                 NaN         NaN                  NaN          NaN   
1                 NaN         NaN                  NaN          NaN   
2                 1.0         NaN                  NaN          NaN   
3                 1.0         NaN                  NaN          NaN   
4                 NaN         NaN                  NaN          NaN   

   YouTubeSimplilearn  YouTubeTheNewBoston  
0                 NaN                  NaN  
1                 NaN                  NaN  
2                 NaN                  NaN  
3                 NaN                  NaN  
4                 NaN                  NaN  

[5 rows x 136 columns]

For the purpose of our analysis, we want to answer questions about a population of new coders that are interested in the subjects we teach. We'd like to know:

* Where are these new coders located.
* What are the locations with the greatest number of new coders.
* How much money new coders are willing to spend on learning.

Before starting to analyze the sample data we have, we need to clarify whether it's representative for our population of interest and it has the right categories of people for our purpose.

In [4]:
sample_summary = data['JobRoleInterest'].value_counts(normalize=True) * 100
print(sample_summary)

Full-Stack Web Developer                                                                                                                                                                                                                                                    11.770595
  Front-End Web Developer                                                                                                                                                                                                                                                    6.435927
  Data Scientist                                                                                                                                                                                                                                                             2.173913
Back-End Web Developer                                                                                                                                                

Based on the result above, people are interested in multiple roles instead of just one. 
Let's check out the frequency table for people who are interested in Web/Mobile Development!

In [5]:
interests_no_nulls = data['JobRoleInterest'].dropna()
multiple_interest = interests_no_nulls.str.split(',')
print(multiple_interest.head(10))

1                            [Full-Stack Web Developer]
2     [  Front-End Web Developer,  Back-End Web Deve...
3     [  Front-End Web Developer,  Full-Stack Web De...
4     [Full-Stack Web Developer,  Information Securi...
6                            [Full-Stack Web Developer]
9     [Full-Stack Web Developer,    Quality Assuranc...
11    [  DevOps / SysAdmin,    Data Scientist,  Info...
13    [Back-End Web Developer,  Full-Stack Web Devel...
14                           [Full-Stack Web Developer]
15                           [Full-Stack Web Developer]
Name: JobRoleInterest, dtype: object


In [6]:
count_of_people_multiple_interest = multiple_interest.apply(lambda x : len(x))
count_of_people_multiple_interest.value_counts(normalize = True).sort_index() * 100

1     31.650458
2     10.883867
3     15.889588
4     15.217391
5     12.042334
6      6.721968
7      3.861556
8      1.759153
9      0.986842
10     0.471968
11     0.185927
12     0.300343
13     0.028604
Name: JobRoleInterest, dtype: float64

In [7]:
web_or_mobile = interests_no_nulls.str.contains(
    'Web Developer|Mobile Developer')
freq_table = web_or_mobile.value_counts(normalize = True) * 100
freq_table

True     86.241419
False    13.758581
Name: JobRoleInterest, dtype: float64

In [8]:
labels = ['Web or mobile development', 'Other subject']
values = [86.24,13.76]
colors = ['#009a6b', '#A9A9A9']
title_text = 'How Popular is Web or Mobile Development<br>Among Participants?'

trace = go.Pie(labels=labels, values=values,
               title=dict(text=title_text, font=dict(size=26)),
               hoverinfo='label+percent', textinfo='value', 
               textfont=dict(size=26),
               showlegend=False,
               opacity=0.7,
               marker=dict(colors=colors, 
                           line=dict(color='#004e36', width=3)))

py.iplot([trace], filename='styled_pie_chart')
#
# DISCLAIMER: 
#
# I am not really a fan of pie chart to convey complex information,
# However! I believe it is still can be effectively used to convey x:y ratio, 
# as such being displayed below.
#
# For more than 2 parameters, I would easily choose bar chart instead pie or donut chart.

Now that we found out that the sample has the right categories of people for our purpose, we can begin analyzing it. We can start with finding out where these new coders are located, and what are the densities (how many coderes there are) for each location.

The data set provides information about the location of each participant at a country level. The `CountryCitizen` variable describes the country of origin for each participant, and the `CountryLive` variable describes what country each participants lives in (which may be different than the origin country).

For our analysis, we'll work with the `CountryLive` variable because we're interested where people actually live at the moment when we run the ads. In other words, we're interested where people are located, not where they were born.

Because the data set provides information at a country level, we can think of each country as an individual market. This means we can frame our goal as finding the two best countries to advertise in.

One indicator of a good market is the number of potential customers — the more potential customers in a market, the better. If our ads manage to convince 10% of the 5000 potential customers in market A to buy our product, then this is better than convincing 100% of the 30 potential customers in market B.

Now, we drop all the rows where participants didn't answer what role they are interested in. Where a participant didn't respond, we can't know for sure what their interests are, so it's better if we leave out this category of participants.


In [9]:
country_live_data = data[data['JobRoleInterest'].notnull()].copy()

absolute_country_live_freq = country_live_data['CountryLive'].value_counts()
relative_country_live_freq = country_live_data['CountryLive'].value_counts(normalize=True)*100

pd.DataFrame(data={'Absolute Frequency' : absolute_country_live_freq,
            'Percentage' : relative_country_live_freq}).head()

Absolute Frequency  Percentage
United States of America                3125   45.700497
India                                    528    7.721556
United Kingdom                           315    4.606610
Canada                                   260    3.802281
Poland                                   131    1.915765

We can continue by figuring out how much money new coders are actually willing to spend on learning. Advertising within markets where most people are only willing to learn for free is extremely unlikely to be profitable for us.

The `MoneyForLearning` column describes in American dollars the amount of money spent by participants from the moment they started coding until the moment they completed the survey. Our company sells subscriptions at a price of $59 per month, and for this reason we're interested in finding out how much money each student spends per month.

It also seems like a good idea to narrow down our analysis to only four countries: the US, India, the United Kingdom, and Canada. Two reasons for this decision are:

* These are the countries having the highest absolute frequencies in our sample, which means we have a decent amount of data for each.
* Our courses are written in English, and English is an official language in all these four countries. The more people that know English, the better our chances to target the right people with our ads.

In [10]:
# Some students answered that they had been learning to code for 0 months
# (it might be that they had just started when they completed the survey). 
# To avoid dividing by 0, replace all the values of 0 with 1.
country_live_data['MonthsProgramming'].replace(0,1, inplace = True)

# We find out the amount of money a student has spent per month
country_live_data['money_per_month'] = country_live_data['MoneyForLearning'] / country_live_data['MonthsProgramming']
country_live_data['money_per_month'].isnull().sum()

# We remove non-null rows
country_live_data = country_live_data[country_live_data['money_per_month'].notnull()]
country_live_data = country_live_data[country_live_data['CountryLive'].notnull()]
country_live_data['CountryLive'].value_counts().head()

United States of America    2933
India                        463
United Kingdom               279
Canada                       240
Poland                       122
Name: CountryLive, dtype: int64

In [11]:
country_live_data_mean = country_live_data.groupby('CountryLive').mean()
country_live_data_mean['money_per_month'][['United States of America', 'India', 'United Kingdom', 'Canada']]

CountryLive
United States of America    227.997996
India                       135.100982
United Kingdom               45.534443
Canada                      113.510961
Name: money_per_month, dtype: float64

The results for the United Kingdom and Canada are surprisingly low relative to the values we see for India. If we considered a few socio-economical metrics (like GDP per capita), we'd intuitively expect people in the UK and Canada to spend more on learning than people in India.

It might be that we don't have have enough representative data for the United Kingdom, Canada, and India, or we have some outliers (maybe coming from wrong survey answers) making the mean too big for India, or too low for the UK and Canada. Or it might be that the results are correct.

## Visualizing Each Country With BoxPlot

In [12]:
top_4 = (country_live_data[country_live_data['CountryLive']
                           .str.contains('United States of America|India|United Kingdom|Canada')]
        )
title_text = 'Money Spent Per Month Per Country<br>(Distributions)'
(top_4[['money_per_month', 'CountryLive']]
     .pivot(columns='CountryLive', values='money_per_month')
     .iplot(kind='box', title=title_text, dimensions=(800,500), boxpoints='suspectedoutliers')
)

In [13]:
# Eliminate extreme outlier
country_live_data = country_live_data[country_live_data['money_per_month'] < 10000]

Now let's recompute the mean values and check the box plots again.

In [14]:
# Recompute mean sum of money spent by students each month
country_live_data_mean  = country_live_data.groupby('CountryLive').mean()
country_live_data_mean ['money_per_month'][['United States of America',
                            'India', 'United Kingdom',
                            'Canada']]

CountryLive
United States of America    155.459187
India                       113.748387
United Kingdom               45.534443
Canada                      113.510961
Name: money_per_month, dtype: float64

In [15]:
top_4 = (country_live_data[country_live_data['CountryLive']
                           .str.contains('United States of America|India|United Kingdom|Canada')]
        )
(top_4[['money_per_month', 'CountryLive']]
     .pivot(columns='CountryLive', values='money_per_month')
     .iplot(kind='box', title=title_text, dimensions=(800,500), boxpoints='suspectedoutliers')
)

We can see a few extreme outliers for India (values over \$2500 per month), but it's unclear whether this is good data or not. Maybe these persons attended several bootcamps, which tend to be very expensive. Let's examine these two data points to see if we can find anything relevant.

In [16]:
# Inspect the extreme outliers for India
india_outliers = top_4[
    (top_4['CountryLive'] == 'India') & 
    (top_4['money_per_month'] >= 2500)]
india_outliers

Age  AttendedBootcamp  BootcampFinish  BootcampLoanYesNo BootcampName  \
1728   24.0               0.0             NaN                NaN          NaN   
1755   20.0               0.0             NaN                NaN          NaN   
7989   28.0               0.0             NaN                NaN          NaN   
8126   22.0               0.0             NaN                NaN          NaN   
15587  27.0               0.0             NaN                NaN          NaN   

       BootcampRecommend  ChildrenNumber                 CityPopulation  \
1728                 NaN             NaN  between 100,000 and 1 million   
1755                 NaN             NaN            more than 1 million   
7989                 NaN             NaN  between 100,000 and 1 million   
8126                 NaN             NaN            more than 1 million   
15587                NaN             NaN            more than 1 million   

       CodeEventConferences  CodeEventDjangoGirls       ...         \
1728                    NaN                   NaN       ...          
1755                    NaN                   NaN       ...          
7989                    1.0                   NaN       ...          
8126                    NaN                   NaN       ...          
15587                   NaN                   NaN       ...          

       YouTubeFunFunFunction  YouTubeGoogleDev  YouTubeLearnCode  \
1728                     NaN               NaN               NaN   
1755                     NaN               NaN               1.0   
7989                     NaN               NaN               NaN   
8126                     NaN               1.0               NaN   
15587                    NaN               1.0               1.0   

       YouTubeLevelUpTuts  YouTubeMIT  YouTubeMozillaHacks  YouTubeOther  \
1728                  NaN         NaN                  NaN           NaN   
1755                  NaN         1.0                  NaN           NaN   
7989                  NaN         NaN                  NaN           NaN   
8126                  NaN         1.0                  NaN           NaN   
15587                 NaN         1.0                  NaN           NaN   

      YouTubeSimplilearn  YouTubeTheNewBoston  money_per_month  
1728                 NaN                  NaN      5000.000000  
1755                 NaN                  NaN      3333.333333  
7989                 NaN                  NaN      5000.000000  
8126                 NaN                  1.0      5000.000000  
15587                NaN                  NaN      4166.666667  

[5 rows x 137 columns]

It seems that neither participant attended a bootcamp. Overall, it's really hard to figure out from the data whether these persons really spent that much money with learning. The actual question of the survey was "Aside from university tuition, about how much money have you spent on learning to code so far (in US dollars)?", so they might have misunderstood and thought university tuition is included. It seems safer to remove these rows.

In [17]:
# Remove the outliers for India
top_4 = top_4.drop(india_outliers.index)

Looking back at the box plot above, we can also see more extreme outliers for the US (values over \$6000 per month). Let's examine these participants in more detail.

In [18]:
# Examine the extreme outliers for the US
us_outliers = top_4[
    (top_4['CountryLive'] == 'United States of America') & 
    (top_4['money_per_month'] >= 6000)]

pd.options.display.max_columns = None
print(us_outliers)

       Age  AttendedBootcamp  BootcampFinish  BootcampLoanYesNo  \
718   26.0               1.0             0.0                0.0   
1222  32.0               1.0             0.0                0.0   
3184  34.0               1.0             1.0                0.0   
7505  26.0               1.0             0.0                1.0   
9778  33.0               1.0             0.0                1.0   

                                BootcampName  BootcampRecommend  \
718   The Coding Boot Camp at UCLA Extension                1.0   
1222                           The Iron Yard                1.0   
3184                          We Can Code IT                1.0   
7505                                  Codeup                0.0   
9778                            Grand Circus                1.0   

      ChildrenNumber                 CityPopulation  CodeEventConferences  \
718              NaN            more than 1 million                   1.0   
1222             NaN  between 100,000 an

Five of the participants attended bootcamp, which justify the large sums of money spent on learning. Interestingly, only one out of five finished the bootcamp. 

Also, the data shows that all five respondents had been programming for no more than three months when they completed the survey. They most likely paid a large sum of money for a bootcamp that was going to last for several months, so the amount of money spent per month is unrealistic and should be significantly lower (because they probably didn't spend anything for the next couple of months after the survey). As a consequence, we'll remove all of these five outliers.

In the next code block, we'll remove respondents that:

* Didn't attend bootcamps.
* Had been programming for three months or less when at the time they completed the survey.

In [19]:
# Remove the respondents who didn't attendent a bootcamp
no_bootcamp = top_4[
    (top_4['CountryLive'] == 'United States of America') & 
    (top_4['money_per_month'] >= 6000) &
    (top_4['AttendedBootcamp'] == 0)
]

top_4 = top_4.drop(no_bootcamp.index)


# Remove the respondents that had been programming for less than 3 months
less_than_3_months = top_4[
    (top_4['CountryLive'] == 'United States of America') & 
    (top_4['money_per_month'] >= 6000) &
    (top_4['MonthsProgramming'] <= 3)
]

top_4 = top_4.drop(less_than_3_months.index)

In [20]:
# Examine the extreme outliers for Canada
canada_outliers = top_4[
    (top_4['CountryLive'] == 'Canada') & 
    (top_4['money_per_month'] > 4500)]

canada_outliers

Age  AttendedBootcamp  BootcampFinish  BootcampLoanYesNo BootcampName  \
13659  24.0               1.0             0.0                0.0      Bloc.io   

       BootcampRecommend  ChildrenNumber       CityPopulation  \
13659                1.0             NaN  more than 1 million   

       CodeEventConferences  CodeEventDjangoGirls  CodeEventFCC  \
13659                   1.0                   NaN           1.0   

       CodeEventGameJam  CodeEventGirlDev  CodeEventHackathons  \
13659               NaN               NaN                  NaN   

       CodeEventMeetup  CodeEventNodeSchool  CodeEventNone CodeEventOther  \
13659              1.0                  NaN            NaN            NaN   

       CodeEventRailsBridge  CodeEventRailsGirls  CodeEventStartUpWknd  \
13659                   NaN                  NaN                   NaN   

       CodeEventWkdBootcamps  CodeEventWomenCode  CodeEventWorkshops  \
13659                    NaN                 NaN                 1.0   

            CommuteTime CountryCitizen CountryLive EmploymentField  \
13659  30 to 44 minutes         Canada      Canada         finance   

      EmploymentFieldOther    EmploymentStatus EmploymentStatusOther  \
13659                  NaN  Employed for wages                   NaN   

       ExpectedEarning  FinanciallySupporting  FirstDevJob Gender GenderOther  \
13659          60000.0                    NaN          NaN   male         NaN   

       HasChildren  HasDebt  HasFinancialDependents  HasHighSpdInternet  \
13659          NaN      1.0                     0.0                 1.0   

       HasHomeMortgage  HasServedInMilitary  HasStudentDebt  HomeMortgageOwe  \
13659              1.0                  0.0             0.0         250000.0   

       HoursLearning                              ID.x  \
13659           10.0  739b584aef0541450c1f713b82025181   

                                   ID.y    Income  IsEthnicMinority  \
13659  28381a455ab25cc2a118d78af44d8749  140000.0               1.0   

       IsReceiveDisabilitiesBenefits  IsSoftwareDev  IsUnderEmployed  \
13659                            1.0            0.0              0.0   

            JobApplyWhen  JobInterestBackEnd  JobInterestDataEngr  \
13659  I haven't decided                 1.0                  NaN   

       JobInterestDataSci  JobInterestDevOps  JobInterestFrontEnd  \
13659                 1.0                NaN                  1.0   

       JobInterestFullStack  JobInterestGameDev  JobInterestInfoSec  \
13659                   1.0                 1.0                 NaN   

       JobInterestMobile JobInterestOther  JobInterestProjMngr  \
13659                1.0              NaN                  1.0   

       JobInterestQAEngr  JobInterestUX                               JobPref  \
13659                NaN            1.0  work for a multinational corporation   

       JobRelocateYesNo                                    JobRoleInterest  \
13659               NaN    Mobile Developer, Full-Stack Web Developer, ...   

      JobWherePref           LanguageAtHome          MaritalStatus  \
13659    from home  Yue (Cantonese) Chinese  single, never married   

       MoneyForLearning  MonthsProgramming   NetworkID         Part1EndTime  \
13659           10000.0                2.0  41c26f2932  2017-03-25 23:23:03   

            Part1StartTime         Part2EndTime       Part2StartTime  \
13659  2017-03-25 23:20:33  2017-03-25 23:24:34  2017-03-25 23:23:06   

       PodcastChangeLog  PodcastCodeNewbie  PodcastCodePen  PodcastDevTea  \
13659               NaN                NaN             NaN            NaN   

       PodcastDotNET  PodcastGiantRobots  PodcastJSAir  PodcastJSJabber  \
13659            NaN                 NaN           NaN              NaN   

       PodcastNone PodcastOther  PodcastProgThrowdown  PodcastRubyRogues  \
13659          1.0          NaN                   NaN                NaN   

       PodcastSEDaily  PodcastSERadio  PodcastShopTal

Here, the situation is similar to some of the US respondents — this participant had been programming for no more than two months when he completed the survey. He seems to have paid a large sum of money in the beginning to enroll in a bootcamp, and then he probably didn't spend anything for the next couple of months after the survey. We'll take the same approach here as for the US and remove this outlier.

In [21]:
# Remove the extreme outliers for Canada
top_4 = top_4.drop(canada_outliers.index)

Let's recompute the mean values and generate the final box plots.

In [22]:
# Recompute mean sum of money spent by students each month
top_4.groupby('CountryLive').mean()['money_per_month']

CountryLive
Canada                       93.065400
India                        65.758763
United Kingdom               45.534443
United States of America    142.654608
Name: money_per_month, dtype: float64

In [23]:
(top_4[['money_per_month', 'CountryLive']]
     .pivot(columns='CountryLive', values='money_per_month')
     .iplot(kind='box', title=title_text, dimensions=(800,500), boxpoints='suspectedoutliers')
)

Considering the results we've found so far, one country we should definitely advertise in is the US. There are a lot of new coders living there and they are willing to pay a good amount of money each month.

We need to choose one more market though.

## Choosing the Two Best Markets

Obviously, one country we should advertise in is the US. Lots of new coders live there and they are willing to pay a good amount of money each month (roughly \\$143).

We sell subscriptions at a price of \\$59 per month, and Canada seems to be the best second choice because people there are willing to pay roughly \\$93 per month, compared to India (\\$66) and the United Kingdom (\\$45).

The data suggests strongly that we shouldn't advertise in the UK, but let's take a second look at India before deciding to choose Canada as our second best choice:

* \\$59 doesn't seem like an expensive sum for people in India since they spend on average \\$66 each month.
* We have almost twice as more potential customers in India than we have in Canada:

In [24]:
# Frequency table for the 'CountryLive' column
top_4['CountryLive'].value_counts(normalize = True) * 100

United States of America    74.967908
India                       11.732991
United Kingdom               7.163030
Canada                       6.136072
Name: CountryLive, dtype: float64

So it's not crystal clear what to choose between Canada and India. Although it seems more tempting to choose Canada, there are good chances that India might actually be a better choice because of the large number of potential customers.

At this point, it seems that we have several options:

1. Advertise in the US, India, and Canada by splitting the advertisement budget in various combinations:

    * 60% for the US, 25% for India, 15% for Canada.
    * 50% for the US, 30% for India, 20% for Canada; etc.

2. Advertise only in the US and India, or the US and Canada. Again, it makes sense to split the advertisement budget unequally. For instance:

    * 70% for the US, and 30% for India.
    * 65% for the US, and 35% for Canada; etc.

3. Advertise only in the US.

At this point, it's probably best to send our analysis to the marketing team and let them use their domain knowledge to decide. They might want to do some extra surveys in India and Canada and then get back to us for analyzing the new survey data.

# Conclusion

In this project, we analyzed survey data from new coders to find the best two markets to advertise in. The only solid conclusion we reached is that the US would be a good market to advertise in.

For the second best market, it wasn't clear-cut what to choose between India and Canada. We decided to send the results to the marketing team so they can use their domain knowledge to take the best decision.